pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go

```diff
@@ -0,0 +1,534 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) reconcileRestoreCompletion() (reconcile.Result, error) {
+	if !r.cr.RestoreCompleted() {
+		// TODO make this more robust only setting it in case all substeps have been completed?
+		// It might be a little bit redundant because the steps are checked during the reconciliation
+		completionTimeUTC := metav1.Time{Time: clock.Now().UTC()}
+		restoreFinished := true
+		r.cr.Status.Completed = &restoreFinished
+		r.cr.Status.CompletionTime = &completionTimeUTC
+		err := r.UpdateResourceStatus(r.cr)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) reconcileRestoreAPIManagerInSharedSecret() (reconcile.Result, error) {
+	desired := r.apiManagerRestore.CreateAPIManagerSharedSecretJob()
+	if desired == nil {
+		return reconcile.Result{}, nil
+	}
+
+	res, err := r.reconcileJob(desired)
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	if r.cr.Status.APIManagerToRestoreRef == nil {
+		secret, err := r.sharedBackupSecret()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if secret == nil {
+			r.Logger().Info("Shared secret '%s' not found. Waiting...", r.apiManagerRestore.SecretToShareName())
+			return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+		}
+		apimanager, err := r.apiManagerFromSharedBackupSecret()
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		r.cr.Status.APIManagerToRestoreRef = &v1.LocalObjectReference{
+			Name: apimanager.Name,
+		}
+		err = r.UpdateResourceStatus(r.cr)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) sharedBackupSecret() (*v1.Secret, error) {
+	secret := &v1.Secret{}
+	err := r.GetResource(types.NamespacedName{Name: r.apiManagerRestore.SecretToShareName(), Namespace: r.cr.Namespace}, secret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	_, ok := secret.Data[backup.APIManagerSerializedBackupFileName]
+	if !ok {
+		return nil, fmt.Errorf("Expected key '%s' in secret '%s' not found", r.apiManagerRestore.SecretToShareName(), backup.APIManagerSerializedBackupFileName)
+	}
+
+	return secret, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) apiManagerFromSharedBackupSecret() (*appsv1alpha1.APIManager, error) {
+	secret, err := r.sharedBackupSecret()
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil {
+		return nil, fmt.Errorf("Secret '%s' not found", r.apiManagerRestore.SecretToShareName())
+	}
+
+	// https://godoc.org/k8s.io/apimachinery/pkg/runtime/serializer#CodecFactory
+	codecFactory := k8sserializer.NewCodecFactory(r.Scheme())
```

Comment:
Why not simply json Marshall/Unmarshall? 

Add a comment explaining why some weird serializer and deserializer is used

Reply:
This is just Kubernetes way of deserializing objects into a runtime object (provided by apimachinery). The own decoder.Decode method in apimachinery contains information about it using the kind, version (from scheme) to deserialize to a runtime object

Reply:
This is comment should be in the code and not lost in github PR